File: tests/scope/zones.c
/* run.config


   OPT: -load-script tests/scope/zones.ml -val -journal-disable
*/

/* bin/viewer.opt -val tests/scope/zones.c */

int T [10];
struct Tstr { int a; int b; } S;
int X,Y,Z;

int simple (int x, int y, int z) {
  y = 3; //no need for y before
  x ++; //not used
  x = y + z;
  return x;
}

int array1 (int x, int y) {
  T[x] = 3;
  T[0] += y;
  return T[0];
}

int struct1 (int x, int y) {
  struct Tstr s;
  s = S; // lose precision : even if we need s.b after, we need S before
  s.a = x;
  s.b += y;
  return s.a;
}

int ctrl1 (int x, int y, int z) {
  int a;
  if (x) {
    a = y;
    goto Lt2; // to keep Lt2
    Lt2 : ;
  }
  else {
    a = z;
  }
  return a;
}

//================================================================

int Xf, Xg, Yf, Yg;

int f (int x, int y, int z) {
  Xf += x;
  Yf = z;
  return x + y;
}

int g (int a, int b, int c) {
  Xg += b;
  Yg = c;
  return a + b;
}

int caller (int cond, int t, int u, int v) {
  int x1 = 0, y1 = 0, z1 = 0, a1 = 0, b1 = 0, c1 = 0;
  int (*pf)(int, int, int) = cond ? &f : &g;
  f(x1, y1, z1);
  g(a1, b1, c1);
  return (*pf)(t, u, v);
}
//================================================================

int main (int x, int y, int z) {
  simple (x, y, z);
  array1 (x, y);
  struct1 (x, y);
  ctrl1 (x, y, z);

  caller (x, x, y, z);

  return 0;
}

File: tests/scope/zones.ml
(* when using toplevel.top :
bin/topleval.top -val tests/scope/zones.c
#directory "cil/src";;
*)

let fmt =  Format.std_formatter;;

(*
let old_debug = Kernel.Debug.get ();;
Kernel.Debug.set 1;; (* to see sid *)
Format.fprintf fmt "@[%a@]" Printer.pp_file ( Ast.get ());;
Kernel.Debug.set old_debug;;
*)

let find_ret kf_name =
  let kf = Globals.Functions.find_by_name kf_name in
  let stmt = Kernel_function.find_return kf in
    Format.printf "Current program point = return in function %s@\n" kf_name;
    stmt, kf
;;

let find_sid sid =
  let stmt, kf = Kernel_function.find_from_sid sid in
    Format.printf "Current program point = before stmt %d in function %a@\n"
      sid Kernel_function.pretty kf;
    stmt, kf
;;

let find_label kf_name lab_name =
  let kf = Globals.Functions.find_by_name kf_name in
  let stmt = !(Kernel_function.find_label kf lab_name) in
    Format.printf "Current program point = label %s in function %s@\n"
      lab_name kf_name;
    stmt, kf

let compute_and_print pp str_data =
  let stmt, kf = pp in
  let lval_term = !Db.Properties.Interp.lval kf stmt str_data in
  let lval = !Db.Properties.Interp.term_lval_to_lval ~result:None lval_term in
  let (_used_stmts, zones) = !Db.Scope.build_zones kf stmt lval in
    Format.printf "Zones for %s at current program point =@.%a\n@\n"
      str_data !Db.Scope.pretty_zones zones
;;

let main _ =
  let pp = find_ret "simple" in
  compute_and_print pp "x";

  let pp = find_ret "array1" in
  compute_and_print pp "T[0]";
  compute_and_print pp "T[1]";
  compute_and_print pp "T[x]";

  let pp = find_ret "struct1" in
  compute_and_print pp "s.a";
  compute_and_print pp "s.b";
  compute_and_print pp "s";

  let pp = find_ret "ctrl1" in
  compute_and_print pp "a";
  let pp = find_label "ctrl1" "Lt2" in
  compute_and_print pp "a";

  let pp = find_ret "caller" in
  compute_and_print pp "Yf"

let () = Db.Main.extend main
